File: packages/arb-avm-go/vm/stack/tuple.go
/*
 * Copyright 2019, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stack

import (
	"bytes"
	"fmt"

	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
)

type Tuple struct {
	stack value.Value
	count int64
}

func NewTuple(stack value.Value) *Tuple {
	ret := &Tuple{stack, 0}
	return ret
}

func (m *Tuple) Equal(yin Stack) (bool, string) {
	// TODO add tuple equal functionality
	return true, ""
}

func (m *Tuple) Clone() Stack {
	return &Tuple{m.stack.Clone(), m.count}
}

func (m *Tuple) Push(v value.Value) {
	m.stack = value.NewTuple2(v, m.stack)
	m.count++
}

func (m *Tuple) PushInt(v value.IntValue) {
	m.Push(v)
}

func (m *Tuple) PushTuple(v value.TupleValue) {
	m.Push(v)
}

func (m *Tuple) PushCodePoint(v value.CodePointValue) {
	m.Push(v)
}

func (m *Tuple) Pop() (value.Value, error) {
	topTuple, ok := m.stack.(value.TupleValue)
	if !ok {
		// Can only occur if there is an internal implementation bug
		panic(fmt.Sprintf("Stack.Pop: Value in Stack was %v instead of a tuple", value.TypeCodeName(m.stack.TypeCode())))
	}
	if topTuple.Len() == 0 {
		return nil, EmptyError{}
	}

	if topTuple.Len() != 2 {
		panic(fmt.Sprintf("Stack.Pop: Value in Stack was tuple of incorrect length %v", topTuple.Len()))
	}
	m.stack, _ = topTuple.GetByInt64(1)
	m.count--
	return topTuple.GetByInt64(0)
}

func (m *Tuple) String() string {
	var buf bytes.Buffer
	buf.WriteString("[")
	s := m.Clone()
	for !s.IsEmpty() {
		val, _ := s.Pop()
		buf.WriteString(fmt.Sprintf("%v", val))
		if !s.IsEmpty() {
			buf.WriteString(", ")
		}
	}
	buf.WriteString("]")
	return buf.String()
}

func (m *Tuple) PopInt() (value.IntValue, error) {
	val, err := m.Pop()
	if err != nil {
		return value.IntValue{}, err
	}
	v, ok := val.(value.IntValue)
	if !ok {
		return v, nil
	}
	return value.IntValue{}, TypeError{"Int", val}
}

func (m *Tuple) PopTuple() (value.TupleValue, error) {
	val, err := m.Pop()
	if err != nil {
		return value.TupleValue{}, err
	}
	v, ok := val.(value.TupleValue)
	if !ok {
		return value.TupleValue{}, TypeError{"Tuple", val}
	}
	return v, nil
}

func (m *Tuple) PopCodePoint() (value.CodePointValue, error) {
	val, err := m.Pop()
	if err != nil {
		return value.CodePointValue{}, err
	}
	v, ok := val.(value.CodePointValue)
	if !ok {
		return value.CodePointValue{}, TypeError{"CodePointValue", val}
	}
	return v, nil
}

func (m *Tuple) IsEmpty() bool {
	topTuple := m.stack.(value.TupleValue)
	return topTuple.Len() == 0
}

func (m *Tuple) Size() int64 {
	return m.stack.Size()
}

func (m *Tuple) Count() int64 {
	return m.count
}

func (m *Tuple) StateValue() value.Value {
	return value.NewHashOnlyValueFromValue(m.stack)
}

func (m *Tuple) ProofValue(stackInfo []byte) value.Value {
	c := m.Clone()
	vals := make([]value.Value, 0, len(stackInfo))
	for range stackInfo {
		val, _ := c.Pop()
		vals = append(vals, val)
	}
	stack := NewTuple(c.StateValue())
	for i := len(stackInfo) - 1; i >= 0; i-- {
		if stackInfo[i] == 1 {
			stack.Push(vals[i].CloneShallow())
		} else {
			stack.Push(value.NewHashOnlyValueFromValue(vals[i]))
		}
	}
	return stack.stack
}

func (m *Tuple) SolidityProofValue(stackInfo []byte) (value.HashOnlyValue, []value.Value) {
	c := m.Clone()
	vals := make([]value.Value, 0, len(stackInfo))
	for i := range stackInfo {
		val, _ := c.Pop()
		if stackInfo[i] == 1 {
			vals = append(vals, val.CloneShallow())
		} else {
			vals = append(vals, value.NewHashOnlyValueFromValue(val))
		}
	}
	return value.NewHashOnlyValueFromValue(c.StateValue()), vals
}

func (m *Tuple) FullyExpandedValue() value.Value {
	return m.stack
}

File: packages/arb-avm-cpp/avm/src/tuple.cpp
/*
 * Copyright 2019, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "avm/tuple.hpp"

#include "bigint_utils.hpp"
#include "util.hpp"

Tuple::Tuple(value val, TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(1)) {
    tpl->data.push_back(std::move(val));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1, value val2, TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(2)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1, value val2, value val3, TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(3)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1, value val2, value val3, value val4, TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(4)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->data.push_back(std::move(val4));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1,
             value val2,
             value val3,
             value val4,
             value val5,
             TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(5)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->data.push_back(std::move(val4));
    tpl->data.push_back(std::move(val5));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1,
             value val2,
             value val3,
             value val4,
             value val5,
             value val6,
             TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(6)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->data.push_back(std::move(val4));
    tpl->data.push_back(std::move(val5));
    tpl->data.push_back(std::move(val6));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1,
             value val2,
             value val3,
             value val4,
             value val5,
             value val6,
             value val7,
             TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(7)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->data.push_back(std::move(val4));
    tpl->data.push_back(std::move(val5));
    tpl->data.push_back(std::move(val6));
    tpl->data.push_back(std::move(val7));
    tpl->cachedHash = calculateHash();
}

Tuple::Tuple(value val1,
             value val2,
             value val3,
             value val4,
             value val5,
             value val6,
             value val7,
             value val8,
             TuplePool* pool)
    : tuplePool(pool), tpl(pool->getResource(8)) {
    tpl->data.push_back(std::move(val1));
    tpl->data.push_back(std::move(val2));
    tpl->data.push_back(std::move(val3));
    tpl->data.push_back(std::move(val4));
    tpl->data.push_back(std::move(val5));
    tpl->data.push_back(std::move(val6));
    tpl->data.push_back(std::move(val7));
    tpl->data.push_back(std::move(val8));
    tpl->cachedHash = calculateHash();
}

void Tuple::marshal(std::vector<unsigned char>& buf) const {
    buf.push_back(TUPLE + tuple_size());
    for (int i = 0; i < tuple_size(); i++) {
        marshal_value(get_element(i), buf);
    }
}

value Tuple::clone_shallow() {
    Tuple tup(tuplePool, tuple_size());
    for (int i = 0; i < tuple_size(); i++) {
        auto valHash = hash(get_element(i));
        tup.set_element(i, valHash);
    }
    return tup;
}

uint256_t Tuple::calculateHash() const {
    std::array<unsigned char, 1 + 8 * 32> tupData;
    auto oit = tupData.begin();
    tupData[0] = TUPLE + tuple_size();
    ++oit;
    for (int i = 0; i < tuple_size(); i++) {
        auto valHash = hash(get_element(i));
        std::array<uint64_t, 4> valHashInts;
        to_big_endian(valHash, valHashInts.begin());
        std::copy(reinterpret_cast<unsigned char*>(valHashInts.data()),
                  reinterpret_cast<unsigned char*>(valHashInts.data()) + 32,
                  oit);
        oit += 32;
    }

    std::array<unsigned char, 32> hashData;
    evm::Keccak_256(tupData.data(), 1 + 32 * tuple_size(), hashData.data());
    return from_big_endian(hashData.begin(), hashData.end());
}

uint256_t zeroHash() {
    std::array<unsigned char, 1> tupData;
    tupData[0] = TUPLE;
    std::array<unsigned char, 32> hashData;
    evm::Keccak_256(tupData.data(), 1, hashData.data());
    return from_big_endian(hashData.begin(), hashData.end());
}

std::ostream& operator<<(std::ostream& os, const Tuple& val) {
    os << "Tuple(";
    for (int i = 0; i < val.tuple_size(); i++) {
        os << val.get_element(i);
        if (i < val.tuple_size() - 1) {
            os << ", ";
        }
    }
    os << ")";
    return os;
}
